[Paladin] legendary updates

OrcinusDrake opened the pull request.

OrcinusDrake committed:
Ashen Hallow Cancelling

OrcinusDrake committed:
[Paladin] Add "ashen_hallow.remains" and "ticking"
expressions

OrcinusDrake committed:
[Paladin] Prot duty bound gavel

OrcinusDrake committed:
[Paladin] Prot gets avenger's shield from
divine resonance

OrcinusDrake committed:
[Paladin] Divine Resonance's spell isn't the same
as divine toll's exactly.
- One way is that DR is affected by Moment of Glory, probably a bug

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -1331,6 +1334,8 @@ struct ashen_hallow_t : public paladin_spell_t
     hd_damage   = new hallowed_discernment_tick_t( p );
     damage_tick = new ashen_hallow_tick_t( p, hd_damage );
     heal_tick   = new ashen_hallow_heal_tick_t( p );
+    cooldown    = p -> cooldowns.ashen_hallow;
+    cooldown -> duration = p -> covenant.venthyr->cooldown();
```
Do you really need that? this should be automagic'd and get_cooldown() will link to the same object as this action's cooldown as long as the action is called the same name as the argument of get_cooldown()

Melekus commented on engine/class_modules/paladin/sc_paladin.cpp:
```diff
@@ -33,6 +33,7 @@ paladin_t::paladin_t( sim_t* sim, util::string_view name, race_e r )
 {
   active_consecration = nullptr;
   active_hallow       = nullptr;
+  active_h_hallow     = nullptr;
```
I'd find a more explicit name because I had to dig through the code to find out what it stores. Maybe active_hallow_heal?

Melekus commented on engine/class_modules/paladin/sc_paladin_protection.cpp:
```diff
@@ -94,16 +94,35 @@ struct avengers_shield_base_t : public paladin_spell_t
+
+  double action_multiplier() const override
+  {
+    // 2021-06-26 Divine resonance damage is increased by Moment of Glory, but
+    // does not consume the buff.
```
If it's not consumed you'll have to adjust or override avenger's shield's execute code to match that. I suggest checking that the action is background=false before decrementing moment of glory

OrcinusDrake replied:
Divine resonance and divine toll (and ringing clarity for that matter) both inherit from avengers_shield_base_t which does not have the moment of glory consumption
Avenger's shield the active ability is its own subclass of that base with the moment of glory implementation (and other "active ability" stuff)

OrcinusDrake replied:
As a python guy I prefer that sub-classing to the way Shield of the Righteous does it with overloaded functions, but idk which is best practice.

Melekus replied:
Makes sense, I completely overlooked that

OrcinusDrake committed:
[Paladin] Redundant setting of the cooldowns
attribute for ashen hallow. Removed.

OrcinusDrake committed:
[Paladin] Renamed internal variables for clarity
active_h_hallow to active_healing_hallow,
and active_hallow to active_damaging_hallow

OrcinusDrake committed:
[Paladin] Make variables a little easier to search

OrcinusDrake committed:
Merge branch 'simulationcraft:shadowlands' into paladin_legendary_updates
